pkgs/dart_mcp_server/lib/src/utils/cli_utils.dart

```diff
@@ -263,17 +262,23 @@ Future<String> defaultCommandForRoot(
     ),
 };
 
-/// Returns whether or not [rootUri] is an allowed root, either exactly matching
-/// or under on of the [knownRoots].
-Root? _findRoot(String rootUri, List<Root> knownRoots) {
-  for (final root in knownRoots) {
-    final knownRootUri = Uri.parse(root.uri);
-    final resolvedRoot = knownRootUri.resolve(rootUri).toString();
-    if (root.uri == resolvedRoot || p.isWithin(root.uri, resolvedRoot)) {
-      return root;
-    }
+/// Returns whether [uri] is under [root].
+///
+/// Relative uris will always be under [root] unless they escape it with `../`.
+bool _isUnderRoot(Root root, String uri) {
```

Comment:
So, this means that "under" always includes the root directory itself.  Is that OK for all the tools?  Would there be some where the operation shouldn't be allowed on the root directory itself?

I can't think of any, but I just wanted to point that out.

Maybe just point out in the docs that it always considers the root itself to be "under" the root.